### PR TITLE
Fix regex in parser

### PR DESCRIPTION
**Pull request by PNieck**
## Description

This PR fixes bug connected with incorrect regex in `parser` library.

## Motivation and Context

Bug-free libraries.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
https://github.com/phoenix-pilot/phoenix-pilot-core/issues/186 
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [x] New test added: (add PR link here). In this PR
- [ ] Tested by hand on: (list targets here).

## Checklist:

- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing linter checks and tests passed.
- [x] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [x] I will merge this PR by myself when appropriate.

### CODE REVIEW

**Review by kemonats — COMMENTED** [review]
A few comments out of this PR.
1. You are using regex inefficiently.
First you check if a line matches a pattern, then you parse a line.
For example, if the line contain a header you do 3 checks and one parse.
It is really necessary to first check the line type and then parse?
2. Checking the line type can be done more simply, for example like this:
```
char *s = line;
while (isspace(*s)) {
    s++;
}
if ((*s == '\0') {
    /* empty line */
}
else if (*s == '#') {
    /* comment line */
}
else if (*s == '@') {
    /* probably header line */
}
else {
    /* probably field line */
}
```
3. As I already wrote because the file format is quite simple ( similar to *.ini files ) you can write this parsing without using regex, it will be faster.

File: libs/parser/parser.c [hunk 1]
```diff
@@ -161,6 +166,21 @@ void parser_free(parser_t *p)
 }
 
 
+static int parser_headerNameCheck(const char *headerName)
+{
+	regmatch_t regmatch;
+
+	if (regexec(&parser_common.headerNameRegex, headerName, 1, &regmatch, 0) == 0) {
```
[the remark this answers is not in the document]
Comment (PNieck) [reply libs/parser/parser.c:173]: It checks if `headerName` is correct - have consistent syntax with docs.

File: libs/parser/parser.c [hunk 2]
```diff
@@ -161,6 +166,21 @@ void parser_free(parser_t *p)
...
+		/* Check if whole string matches regex */
+		if (regmatch.rm_so == 0 && regmatch.rm_eo == strlen(headerName)) {
+			return 0;
+		}
+	}
+
+	return -1;
```
Comment (kemonats) [inline libs/parser/parser.c:180]: Change pattern in line `445` to `"^" WORD "$"` and
```suggestion
	if (regexec(&parser_common.headerNameRegex, headerName, 1, &regmatch, 0) == 0) {
		return 0;
	}

	return -1;
```
Missing:
```
	regfree(&parser_common.headerNameRegex);
```
in `parser_regexFree(void)`

File: libs/parser/parser.c [hunk 3]
```diff
@@ -31,11 +29,14 @@
...
-#define HEADER_PATTERN        "^[[:space:]]*@(" WORD "+)[[:space:]]*(#.*)?[[:space:]]*$"
+#define WORD "([A-Za-z0-9_]+)"
+
+#define HEADER_PATTERN        "^[[:space:]]*@" WORD "[[:space:]]*(#.*)?[[:space:]]*$"
 #define HEADER_SUBEXPRESSIONS 2
 
-#define FIELD_PATTERN        "^[[:space:]]*(" WORD "+)[ =]+(" WORD "+)([[:space:]]+#.*)?[[:space:]]*$"
-#define FIELD_SUBEXPRESSIONS 3
+#define FIELD_VALUE          "([A-Za-z0-9_,.\\+\\-]+)"
+#define FIELD_PATTERN        "^[[:space:]]*" WORD "( +| *= *)" FIELD_VALUE "([[:space:]]+#.*)?[[:space:]]*$"
```
[the remark this answers is not in the document]
Comment (PNieck) [reply libs/parser/parser.c:38]: No, `=` is also possible

[the remark this answers is not in the document]
Comment (kemonats) [reply libs/parser/parser.c]: What I meant was whether there could be tab characters in addition to spaces.

**Comment by kemonats** [comment]
Remove the line:
```
#define COMMENT_PATTERN "^[[:space:]]*#.*$|^[[:space:]]*$"
```
and line containing variable `commentRegex`

**Comment by PNieck** [comment]
@kemonats  Sorry for not responding for so long, but I was working on another project. Now I am back in Phoenix-Pilot

File: libs/parser/parser.c [hunk 4]
```diff
@@ -161,6 +167,18 @@ void parser_free(parser_t *p)
 }
 
 
+static int parser_headerNameCheck(const char *headerName)
+{
+	regmatch_t regmatch;
+
+	if (regexec(&parser_common.headerNameRegex, headerName, 1, &regmatch, 0) == 0) {
```
[the remark this answers is not in the document]
Comment (PNieck) [reply libs/parser/parser.c:174]: Yes, this will be better

File: libs/parser/parser.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (PNieck) [reply libs/parser/parser.c]: Yes only spces

File: libs/parser/parser.c [hunk 5]
```diff
@@ -161,6 +164,16 @@ void parser_free(parser_t *p)
 }
 
 
+static int parser_headerNameCheck(const char *headerName)
```
Comment (niewim19) [inline libs/parser/parser.c:167]: Provide a short description for this function as with the other, please.

**Review by niewim19 — COMMENTED** [review]
Remark not related to this PR: make `parser_lineTypeGet` static, please. Just to be consistent.

**Comment by PNieck** [comment]
> Remark not related to this PR: make `parser_lineTypeGet` static, please. Just to be consistent.

Done

**Comment by PNieck** [comment]
> A few comments out of this PR.
> 
>     1. You are using regex inefficiently.
>        First you check if a line matches a pattern, then you parse a line.
>        For example, if the line contain a header you do 3 checks and one parse.
>        It is really necessary to first check the line type and then parse?
> 
>     2. Checking the line type can be done more simply, for example like this:
> 
> 
> ```
> char *s = line;
> while (isspace(*s)) {
>     s++;
> }
> if ((*s == '\0') {
>     /* empty line */
> }
> else if (*s == '#') {
>     /* comment line */
> }
> else if (*s == '@') {
>     /* probably header line */
> }
> else {
>     /* probably field line */
> }
> ```
> 
>     3. As I already wrote because the file format is quite simple ( similar to *.ini files ) you can write this parsing without using regex, it will be faster.

@kemonats I see that we can parse file without using regexes 

> A few comments out of this PR.
> 
>     1. You are using regex inefficiently.
>        First you check if a line matches a pattern, then you parse a line.
>        For example, if the line contain a header you do 3 checks and one parse.
>        It is really necessary to first check the line type and then parse?
> 
>     2. Checking the line type can be done more simply, for example like this:
> 
> 
> ```
> char *s = line;
> while (isspace(*s)) {
>     s++;
> }
> if ((*s == '\0') {
>     /* empty line */
> }
> else if (*s == '#') {
>     /* comment line */
> }
> else if (*s == '@') {
>     /* probably header line */
> }
> else {
>     /* probably field line */
> }
> ```
> 
>     3. As I already wrote because the file format is quite simple ( similar to *.ini files ) you can write this parsing without using regex, it will be faster.

@kemonats I see that we can parse files without using regex library. I have added issue, which describes this and I am going to add appropriate PR. In this PR lets just fix parser.

https://github.com/phoenix-pilot/phoenix-pilot-core/issues/197

**Review by niewim19 — DISMISSED** [review]
Of course this can be done without regexes. Its good that its a library so we can work on it in the future for simplification without changing the front-end of it.

LGTM

**Comment by PNieck** [comment]
I have removed the last commit which ignored one of the parser tests. Now problem was solved and test can be active again.
https://github.com/phoenix-rtos/phoenix-rtos-project/issues/628

**Review by niewim19 — APPROVED** [review]
LGTM

**Pull request merged**